### DISCUSSION TITLE
Return validator messages in custom Rule

### DISCUSSION THREAD

**Discussion by slisarenko** [laravel/framework#45997] [Q&A]
A problem that I probably didn't know about when switching to PHP 8+: Very often, in the rules I returned a message like this:
```
public function message(): array
{
    return [$this->validator->errors()->messages()];
} 
```
When using PHP 7.4 - this isn't a problem, but not for PHP 8+

Since looking "deeper" into how the laravel framework forms messages, we get an error in the replaceAttributePlaceholder method of the FormatsMessages class:
```
/**
 * Replace the :attribute placeholder in the given message.
 *
 * @param  string  $message
 * @param  string  $value
 * @return string
 */
protected function replaceAttributePlaceholder($message, $value)
{
    return str_replace(
        [':attribute', ':ATTRIBUTE', ':Attribute'],
        [$value, Str::upper($value), Str::ucfirst($value)],
        $message
    );
}
```
And indeed, if open any editor and run the same code, but for two different versions, we'll get:
![Знімок екрана 2023-01-30 о 19 48 14](https://user-images.githubusercontent.com/55946144/217245829-22aa153c-9927-41f8-80ab-4f927e5d78e3.png)
**Importantly:**
If you return the message like this:

```
public function message(): array
{
    return $this->validator->errors()->messages();
}
```
We will avoid the error, but accordingly, the format of the message will be different - this doesn't suit me, and the format of the message should remain the same. 
Does anyone have any ideas on how to save the format and fix the error?

**Comment by Kristories** [comment]
When using PHP 8+, you may encounter errors when trying to return validator messages in custom rules in the framework. This can be due to a mismatch in syntax between the older PHP 7.4 version and the newer PHP 8+ version. 

In order to fix this error, you must modify the code to account for the syntax differences. Specifically, you need to change the return statement in the message function from: 

```
public function message(): array
{
    return [$this->validator->errors()->messages()];
}
```

to 

```
public function message(): array
{
    return $this->validator->errors()->messages();
}
```

This will ensure that the format of the message remains the same and the error is resolved. Additionally, you should also check and update any other code that might be affected by the change in syntax, such as the `replaceAttributePlaceholder` method of the `FormatsMessages` class. By making the appropriate changes, you should be able to ensure that your code works correctly with PHP 8+.

**Comment by slisarenko** [comment]
@Kristories  You say that the proposed changes will not affect the message format, but that is not the case. 
I see it works differently and woauld like to give an example.
I validate content that can have two sides, **right** and **left**, both sides must be attached, JSON:
```
{
    "data": {
        "content": {}
    }
}
```
We get an error because there are no sides in the example and return a message:
```
PHP 7.4: 

  public function message(): array
  {
      return [$this->validator->errors()->messages()];
  }

Result of the function:

  {
    "errors": {
      "data.content": [
        "The data.content must have at least 2 items.",
        {
          "left": [
            "The left field is required."
          ],
          "right": [
            "The right field is required."
          ]
        }
      ]
    }
  }    
```
```
PHP 8.1: 

  public function message(): array
  {
      return $this->validator->errors()->messages();
  }

Result of the function:

  {
    "errors": {
      "data.content": [
        "The data.content must have at least 2 items.",
        [
          "The left field is required."
        ],
        [
          "The right field is required."
        ]
      ]
    }
  }
```
That's why I have a question to which there is no answer yet, how can you save the format for PHP 8.1 that we have for PHP 7.4 by performing minimal fixes ...
